### PR TITLE
LICM: add an optimization to move multiple loads and stores from/to the same memory location out of a loop.

### PR DESCRIPTION
**Pull request by eeckstein**
This is a combination of load hoisting and store sinking, e.g.
```
  preheader:
    br header_block
  header_block:
    %x = load %not_aliased_addr
    // use %x and define %y
    store %y to %not_aliased_addr
    ...
  exit_block:
```
is transformed to:
```
  preheader:
    %x = load %not_aliased_addr
    br header_block
  header_block:
    // use %x and define %y
    ...
  exit_block:
    store %y to %not_aliased_addr
```

This optimization is important to optimize inout arguments, especially with COW support in SIL: it relies on values being in SSA values rather than in memory.

This is the second try to land https://github.com/apple/swift/pull/27849. It includes the fix which broke the ASAN bot.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/SILOptimizer/LoopTransforms/LICM.cpp [hunk 1]
```diff
@@ -737,14 +819,195 @@ bool LoopTreeOptimization::optimizeLoop(
...
+  // If there are multiple stores in a block, only the last one counts.
+  Optional<SILLocation> loc;
+  for (SILInstruction *I : LoadsAndStores) {
+    if (auto *SI = isStoreToAddr(I, addr)) {
+      loc = SI->getLoc();
+
+      // If a store just stores the loaded value, bail. The operand (= the load)
+      // will be removed later, so it cannot be used as available value.
+      // This corner case is suprisingly hard to handle, so we just give up.
+      if (isLoadFromAddr(dyn_cast<LoadInst>(SI->getSrc()), addr))
```
Comment (atrick) [inline lib/SILOptimizer/LoopTransforms/LICM.cpp]: I guess this `isLoadFromAddr` was a bug that you found after the first implementation, but I really don't get it. Just remove the store if it's not changing the available value! The SSA updater should get it right if it doesn't see this store as a def.

**Pull request merged**